api/server/sdk/cloud_backup.go

```diff
@@ -0,0 +1,252 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package sdk
+
+import (
+	"context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/libopenstorage/openstorage/api"
+	"github.com/libopenstorage/openstorage/volume"
+)
+
+// CloudBackupServer is an implementation of the gRPC OpenStorageCloudBackup interface
+type CloudBackupServer struct {
+	driver volume.VolumeDriver
+}
+
+// Create creates a backup for a volume
+func (s *CloudBackupServer) Create(
+	ctx context.Context,
+	req *api.SdkCloudBackupCreateRequest,
+) (*api.SdkCloudBackupCreateResponse, error) {
+
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply a volume id")
+	} else if len(req.GetCredentialUuid()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply credential uuid")
+	}
+
+	// Create the backup
+	if err := s.driver.CloudBackupCreate(&api.CloudBackupCreateRequest{
+		VolumeID:       req.GetVolumeId(),
+		CredentialUUID: req.GetCredentialUuid(),
+		Full:           req.GetFull(),
+	}); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to create backup: %v", err)
+	}
+
+	return &api.SdkCloudBackupCreateResponse{
+		// THIS WILL BE FIXED BY #### where Create will return an ID
+		BackupId: "TBD",
+	}, nil
+}
+
+// Restore a backup
+func (s *CloudBackupServer) Restore(
+	ctx context.Context,
+	req *api.SdkCloudBackupRestoreRequest,
+) (*api.SdkCloudBackupRestoreResponse, error) {
+
+	if len(req.GetBackupId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must provide backup id")
+	} else if len(req.GetCredentialUuid()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must provide credential uuid")
+	}
+
+	r, err := s.driver.CloudBackupRestore(&api.CloudBackupRestoreRequest{
+		ID:                req.GetBackupId(),
+		RestoreVolumeName: req.GetRestoreVolumeName(),
+		CredentialUUID:    req.GetCredentialUuid(),
+		NodeID:            req.GetNodeId(),
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to restore backup: %v", err)
+	}
+
+	return &api.SdkCloudBackupRestoreResponse{
+		RestoreVolumeId: r.RestoreVolumeID,
+	}, nil
+
+}
+
+// Delete deletes a backup
+func (s *CloudBackupServer) Delete(
+	ctx context.Context,
+	req *api.SdkCloudBackupDeleteRequest,
+) (*api.SdkCloudBackupDeleteResponse, error) {
+
+	if len(req.GetBackupId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must provide backup id")
+	} else if len(req.GetCredentialUuid()) == 0 {
```

Comment:
Correct me if i am wrong
pxctl doesnt place a mandatory requirement on providing cred uuid if there is only one credential present in the cluster.

Shouldn't it be the same for SDK ?

Reply:
on second thoughts , I might disagree with my own statement.
but food for thought.